### ISSUE TITLE
Where to apply an Absolute/Relative distinction?

### ISSUE THREAD

**Issue by keilw** [unitsofmeasurement/unit-api#140]
Based on input from https://reference.wolfram.com/language/tutorial/TemperatureUnits.html
or https://www.quora.com/What-are-absolute-units (the latter contradicts itself quite a bit because after saying "**Measurement units** can be either absolute or relative… " it continues like "However, **physical measurements** are often relative. ") Let's figure out, if `Unit` or `Quantity` are the best place for such `Absolute/Relative` indicator?

It is drafted in `Quantity` (mainly to find out the payload it adds on top of other elements like the measurement level) but e.g. the [Mathematica/Wolfram examples](https://reference.wolfram.com/language/tutorial/TemperatureUnits.html) are pretty clear about using specialized `Unit` definitions like `Kelvin` vs. `KelvinDifference` for that.

Not entirely sure, which is best, something like an attribute on `Unit` which could allow `Units.KELVIN` vs. `Units.KELVIN_DIFFERENCE` (I recall @desruisseaux mentioned that option once in an earlier thread) or an attribute on `Quantity`?

Also how should it be defined either **statically** or **dynamically** and how does it affect unit and quantity operations?

Needs #95 
Needs #131

**Comment by keilw** [comment]
Is it possible to answer this?

**Comment by keilw** [comment]
A) Define the absolute/relative indicator in Quantity

**Comment by andi-huber** [comment]
As demonstrated with [1], the Mathematica way of handling this seems almost trivial to implement.
Note: That would be Option c), where the distinction (absolute vs relative) is made by using different units.

That's not my vote, because I'm not sure how well the other options could be implemented. But I have a feeling, that option c) could be the most simple one.

[1] https://github.com/unitsofmeasurement/indriya/pull/164

**Comment by andi-huber** [comment]
We also need to carefully consider impact on Unit formatting and parsing. E.g. currently I'm using this JAXB-Adapter to allow for Quantities to be marshaled/unmarshaled ...

```
public static final class QuantityAdapter extends XmlAdapter<String, Quantity<?>>{
	
	private final UnitFormat format = SimpleUnitFormat.getInstance();

	public Quantity<?> unmarshal(String v) throws Exception {
		final String unitLiteral = Strings.substringEmbeddedBy(v, "[", "]");
		final String valueLiteral = Strings.substringEmbeddedBy(v, null, "[");
		final double value = Double.parseDouble(valueLiteral);
		final Unit<?> unit = format.parse(unitLiteral);
		return Quantities.getQuantity(value, unit);
	}

	public String marshal(Quantity<?> v) throws Exception {
		
		String formattedUnit = format
				.format(v.getUnit(), new StringBuilder()).toString();
		
		return v.getValue().doubleValue() + "[" + formattedUnit + "]";
	}

}
```
Having different Units that yield the same format, would break this code. So even if we choose option c) an additional flag indicating 'absolute' vs 'relative' would be required somewhere.

**Comment by desruisseaux** [comment]
Absolute/relative indicator in `Quantity` is fine. Note however that it has the following impact: Since `Unit.getConverterTo(Unit)` does not know if we want to convert absolute or relative quantities, we may need another mechanism to distinguish the two cases at `UnitConverter` level. One possible approach is to add a new method in `UnitConverter` interface, which would apply on relative values:

```
double deltaConvert(double);
```

This is the same approach than `java.awt.geom.AffineTransform`, which has a `deltaTransform` method ([javadoc here](https://docs.oracle.com/en/java/javase/11/docs/api/java.desktop/java/awt/geom/AffineTransform.html#deltaTransform(java.awt.geom.Point2D,java.awt.geom.Point2D))).

**Comment by keilw** [comment]
@andi-huber Thanks a lot for demonstrating this. 
I thought using different units would rather be **option B**.

**Comment by dautelle** [comment]
hank you andi-huber, the Wolfram article made me change my mind. It is elegant and makes senses mathematically (indeed Wolfram is a reference for that matter).
For the user, it is quite simple, there can be a property in Unit (e.g. isDifference()) and a few additional units (e.g. CELSIUS_DIFF, FAHRENHEIT_DIFF). Of course, some quantity operations will not be allowed (e.g. adding quantity in CELSIUS and quantity in FAHRENHEIT not allowed, but adding quantity in CELSIUS_DIFF and quantity in FAHRENHEIT_DIFF allowed).

**Comment by keilw** [comment]
@andi-huber @dautelle Thanks for helping with the decision making and showing how the Wolfram calculations may best work here. A property (not really biassed towards `isAbsolute()`, `isRelative()` or `isDifference()`) on `Unit` instead of `Quantity` looks fine in this case. As the `Units` class also implements a `SystemOfUnits`, does it make sense to add those `_DIFF` or _`DIFFERENCE `units there? Keep in mind, `SI` in https://github.com/unitsofmeasurement/si-units extends this class. It is not written in stone, but adds reusability. Thus is it better to define those units in the `Units` class or add a new `SystemOfUnits` implementation like `UnitsDifference`?

**Comment by desruisseaux** [comment]
I'm not convinced by the Wolfram examples. They are mathematically inconsistent (they said themselves that their operations are not commutative) and their approach would bring new problems in JSR-363 context, discussed below. The Wolfram example stating 3°F + 2°C as illegal but 3°F + 2°F = 5°F as legal is inconsistent: for 3°F + 2°F they implicitly handle at least one °F value as a temperature difference despite the two units declared as "DegreesFahrenheit" (and not "DegreesFahrenheitDifference"). In addition of bringing the mathematical problems described in #95, this is inconsistent with the distinction they make between "DegreesFahrenheit" and "DegreesFahrenheitDifference" in other examples, and is also inconsistent with the fact that they do not handle any value as a temperature difference in 3°F + 2°C.

The [examples they provide](https://reference.wolfram.com/language/tutorial/TemperatureUnits.html) can be practically the same with absolute/difference flag in `Quantity` instead of `Unit`. For example instead of (in Mathematica syntax):

```
Quantity[3., "DegreesCelciusDifference"]
```

We would have (in Java syntax)

```
Quantities.create(3., Units.CELSIUS, someFlagForDifference)
```

Having "DegreesCelciusDifference" and "DegreesCelcius" as two different `Unit` instances raises new problems in JSR 363 context. Do we create a new `TemperatureDifference` quantity in addition of `Temperature`?

* If we do not define new `Quantity` subtypes, then the assumption that `Unit.getConverterTo(Unit)` do not fail if the parameterized `Quantity` is the same does not hold anymore, because "DegreesCelciusDifference" would have the same parameterized `Quantity` than "DegreesCelcius" and still be unconvertible.
* If we define new `Quantity` subtype, then users may need to duplicate formulas involving temperature. For example E = 1.5*kT is valid for both an absolute temperature in Kelvin and a temperature difference, in which case E is a difference of energy. User can no longer use the same formula for both. Furthermore we would have to duplicate other `Quantity` subtypes, not just temperature. Example: density sigma-t, pH…

**Comment by keilw** [comment]
Based on @andi-huber's sample it would be` Quantities.create(3., Units.CELSIUS_DIFFERENCE)` or similar, The flag if any was needed (@dautelle hinted it might help) would be on the `Unit`. 
Having separate `Unit` instances with different properties, underneath, that seems relatively uninvasive, but having to introduce `Temperature` vs. `TemperatureDifference`, that would totally blow up the API (even if it was only done for `Temperature`) and introduce all sorts of compatibility problems.

Btw, what about `LevelOfMeasurement`, is there any use for that at all?

**Comment by keilw** [comment]
`LevelOfMeasurement` on `Unit` contradicts everything that was discussed and concluded earlier in https://github.com/unitsofmeasurement/unit-api/issues/131

While it currently sits on `Quantity `and (if that needs to explicitly be set remains to be seen) can be set via calls to `Quantities.create(3., Units.CELSIUS, INTERVAL)` or similar.
Unless we end up with `CELSIUS_INTERVAL` after all (the on with `INTERVAL` set by default) where would it be set on a `Unit`? Claiming that Mathematica, a system that is considered the de facto standard for these kinds of mathematical calculations is wrong or insufficient seems a bit bold.
Sure, there are several approaches to many things especially in the JDK itself (**Date/Time**, how to deal with **nullness**, Jigsaw or the multi-release JARs) and you find wide-spread criticism of most in one form or another. As a temporary precondition, I gave an` isAbsolute()` flag to both `Unit `and `Quantity`. @andi-huber already demonstrated an approach that solves most of them.

**Comment by dautelle** [comment]
What Wolfram shows is that to be mathematically consistent some operations may be forbidden.
This can be done at the Quantity level or at the Unit level. 
The problem occurs with offset units (e.g. Temperature), no need to create a new Quantity class (e.g. TemperatureDifference). It can be resolved elegantly at the Unit level.
For exemple by adding the "add" and "subtract" operation to Unit returning themselves except for offset units such as CELSIUS for which 'add' raises an exception and 'subtract' returns CELSIUS_DIFFERENCE (note: KELVIN will inherit the default behaviour of returning itself for both operations since it is not an offset unit).

**Comment by andi-huber** [comment]
I believe Martin aims for higher goals, but as a user of UOM I'm satisfied with what Mathematica has to offer. If certain operations are not allowed, at least if I understand why, than I'm fine with it. Also Wolfram's TemperatureUnits guide is short and easy to comprehend.

The discussion whether or not the math with Mathematica is 'consistent', in my opinion depends on how one defines 'consistent'. I do have a feeling, that avoiding surprises for the UOM user is more important.

So, regarding the big picture, it appears to me that we have 2 competing ideas:
1) keep it simple, disallow ambiguous calculations, do it the Mathematica way
2) implement Martin's requirement for 'physical consistency' of arithmetic, but also introduce/allow calculations that might surprise users

As a user, I have no strong preference for 1) or 2), all I'm saying is that for me 1) is sufficient.

... regarding exceptions: One way to deal with operations, that are not allowed, is to return a special class like `NonEvaluatedQuantity`, analog to `Double.NaN` when Number calculation yields 'not a number' result

**Comment by andi-huber** [comment]
Also note, with Mathematica ...
1) `10m + 5ft` -> `37.8084 ft`
2) `UnitConvert[Quantity[3., "KelvinsDifference"], "Kelvins"]` -> 'KelvinsDifference and Kelvins are incompatible units'
3) there is no "MetersDifference"

**Comment by andi-huber** [comment]
@desruisseaux, thanks for the examples. These could be key to understanding, what the discussion is all about.

I suggest we translate your examples into Mathematica syntax for having concrete cases we can focus the discussion. Otherwise I feel the subject is too abstract to comprehend. (At least for me.)

Best outcome would be, that we all come to an agreement, what we mean by 'consistent' or 'surprising' and/or identify cases, where we agree Mathematica does something 'wrong' and we can do better.

**Comment by desruisseaux** [comment]
There is my try:

```
k = Quantity[1.3806504E−23, "JoulePerKelvin"]
T = Quantity[10, "DegreeCelsius"]
```

I'm not sure what would be Mathematica behavior with `1.5 * k * T` - maybe it would raises an error. #95 proposal is to convert T to Kelvin automatically. Let suppose that Mathematica can do that too, then the following produce different results in Mathematica:

* `E = 1.5 * k * T` if executed as `1.5 * k * UnitConvert[T, "Kelvin"]`
* `E = 1.5 * T * k` if executed as `UnitConvert[1.5 * T, "Kelvin"] * k`

See "Multiplication and unit conversions are non-commutative for absolute temperatures" section in their [examples](https://reference.wolfram.com/language/tutorial/TemperatureUnits.html). See also the following example:

```
UnitConvert[Quantity[3., "DegreesCelsius"] *18.2, "Kelvins"] = 327.75 K
```

That example computed 3°C \* 18.2 = 54.6°C, then added 273.15 K. Note that they did that despite the unit being "DegreesCelsius", not "DegreesCelsiusDifference" (see also the other inconsistencies reported in above comment). Their calculation is in contradiction with [other source](https://rechneronline.de/temperatur/temperature.php) saying that "The common belief, that twice 10°C (or °F) equals 20°C (or °F), is wrong."

**Comment by keilw** [comment]
@andi-huber Thanks for the input also on other quantities. 

@desruisseaux 
"...- which is also consistent with thermodynamic laws". This shows, the entire discussion is about an extremely narrow case. Everywhere else operations like `10m + 5ft -> 37.8084 ft` (the only difference is, JSR 363 or rather most of its implementations interpret that in the first unit `m`, not the second one, but that is a detail and either one makes perfect sense)

We should focus on concrete solutions for these narrow cases without breaking everything for 99% of all other cases or making things overly complicated for the common use case.

Java won't allow us to identify the `Q` in `Quantity<Q> add(Quantity<Q> addend);` Other languages like Kotlin, C# or F# may be better but unless the JDK team and Oracle thinks this was an important feature to add to Java 13 or 20 we can't do anything about it. 
So looking at a `Quantity` type like `Temperature` won't work, unless we gave it some kind of meta-information, but then this would have to be done everyhere. We could "blacklist" or specially treat certain units, this is done in a limited form e.g. for `GRAM` vs. `KILOGRAM`, etc. when formatting and parsing.  In the end it does not matter much, if `LevelOfMeasurement` was applied to `Unit` or `Quantity`, it is a property we could more easily use to prevent certain operations than on the quantity type itself.

**Comment by keilw** [comment]
@desruisseaux We currently have that with `Quantities.create(5, Units.CELSIUS, INTERVAL)` using the `LevelOfMeasurement`. If everyone agrees we must use an Absolute/Relative flag instead, why not, it's already there, but then do we still need the `LevelOfMeasurement` at all or can we just use isAbsolute() or `isDifference()` or (if the other enum is worthless) then even an `enum` with a reasonable name might work.
The `enum` is there, I even started applying it to @andi-huber's PoC based on the Mathematica case, so please let's use that enum for now (RATIO vs. INTERVAL) and demonstrate that it solves all the problems where needed and has no negative side-effects or adds too much complication for all the others.

**Comment by andi-huber** [comment]
> 
> 
> There is my try:
> 
> ```
> k = Quantity[1.3806504E−23, "JoulePerKelvin"]
> T = Quantity[10, "DegreeCelsius"]
> ```
> 
For the sake of demonstration I redefined k ...
k = Quantity[1, "Joule"]/Quantity[1, "Kelvins"];
tc = Quantity[10, "DegreesCelsius"];

Correct answer to calculating `1.5 * toKelvin(tc) * k` is `424.725 J`.
With Mathematica (correct) ...
`1.5  * UnitConvert[tc, "Kelvins"] * k` -> `424.725 J`

> I'm not sure what would be Mathematica behavior with `1.5 * k * T` - maybe it would raises an error. #95 proposal is to convert T to Kelvin automatically. Let suppose that Mathematica can do that too, then the following produce different results in Mathematica:
> 
>     * `E = 1.5 * k * T` if executed as `1.5 * k * UnitConvert[T, "Kelvin"]`
> 
>     * `E = 1.5 * T * k` if executed as `UnitConvert[1.5 * T, "Kelvin"] * k`
Surprise with Mathematica (both wrong) ...
```
(1.5 * tc ) * k -> 4112.25 J
(1.5 * k ) * tc -> 4112.25 J
```
Root reason for this seems that Mathematica calculates ...
`Quantity[10, "DegreesCelsius"] / Quantity[2, "Kelvins"] -> 5`
which is a surprise to me.
> 
> See "Multiplication and unit conversions are non-commutative for absolute temperatures" section in their [examples](https://reference.wolfram.com/language/tutorial/TemperatureUnits.html). See also the following example:
> 
> ```
> UnitConvert[Quantity[3., "DegreesCelsius"] *18.2, "Kelvins"] = 327.75 K
> ```
This does not surprise me, as long as `a + a = 2 * a` (given a is a quantity, regardless of unit).
eg ...
```
Quantity[3, "DegreesCelsius"] + Quantity[3, "DegreesCelsius"]  -> 6°C
2 * Quantity[3, "DegreesCelsius"] -> 6°C
```
agreed?
> 
> That example computed 3°C * 18.2 = 54.6°C, then added 273.15 K. Note that they did that despite the unit being "DegreesCelsius", not "DegreesCelsiusDifference" (see also the other inconsistencies reported in above comment). Their calculation is in contradiction with [other source](https://rechneronline.de/temperatur/temperature.php) saying that "The common belief, that twice 10°C (or °F) equals 20°C (or °F), is wrong."

I'd like to emphasize that 
`Quantity[10, "DegreesCelsius"] / Quantity[2, "Kelvins"] -> 5`
certainly is something, we can do better. This calculation should not be allowed, in Mathematica's context.

**Comment by desruisseaux** [comment]
> This does not surprise me, as long as a + a = 2 * a (given a is a quantity, regardless of unit).

Yes, we agree that this relationship must be kept. However the result depends if _a_ is a temperature or a temperature difference: 2°C + 2°C = 4°C in the later case, while 2°C + 2°C = 277.15 °C in the former case. Mathematica surprise in my opinion is to have two different units for distinguishing those two cases ("DegreeCelsius" and "DegreeCelsiusDifference"), but still ignore the user unit ("DegreeCelsius") by silently replacing it with another unit ("DegreeCelsiusDifference") which does not have the same meaning.

**Comment by andi-huber** [comment]
@desruisseaux, I remember we discussed this earlier this year: Assume, you can leave the physical interpretation of any quantity aside, than the calculation results should not be surprising. Conceptually this is just one possible (yet simple) way to do it. The consequences being, that certain operations are not (or should not be) allowed, as well as unit-conversions are not commuting with other transformations. 

To recap on
> Quantity[10, "DegreesCelsius"] / Quantity[2, "Kelvins"] -> 5

I do accept, that with UOM, we do things differently, such that we bring in the physical interpretation of Quantities (convert to Kelvin before calculation). I also do accept, that as a consequence we can not just copy Mathematica's way.

So at least that seems clear to me now. I hope that's also clear to everyone else.

**Comment by andi-huber** [comment]
Question to you guys, suppose one needs to calculate the difference quantity of 2 absolute temperatures, how to do this with UOM? (Let's for a moment not worry about the naming ABSOLUTE/DIFFERENCE)
```
t1 = Quantities.create(5., Units.CELSIUS, ABSOLUTE);
t2 = Quantities.create(3., Units.CELSIUS, ABSOLUTE);
deltaT = ??? // t1 - t2 = 2°C (DIFFERENCE)
```
I guess `t1.subtract(t2)` should be the same as `t1.add(t2.multiply(-1))` which is an absolute temperature, so that does not work.

**Comment by keilw** [comment]
Would we judge by both the `Unit` (because `Temperature` is not so easy to grasp at runtime) and the absolute/relative distinction?

**Comment by dautelle** [comment]
t1 = Quantities.create(5., Units.CELSIUS, ABSOLUTE);
t2 = Quantities.create(3., Units.CELSIUS, ABSOLUTE);
deltaT = ??? // t1 - t2 = 2°C (DIFFERENCE)
Either it returns a Quantity with the DIFFERENCE flag set or if it returns a general Quantity stated in CELSIUS_DIFFERENCE

My personal feeling is that a quantity is an "absolute amount" (conceptually), but units can be offset  (could even be more complicated, e.g. Decibels units based on Log function). 
For the cornercase of Température, the simplest solution would be CELSIUS_DIFFERENCE and FAHRENHEIT_DIFFERENCE units (with Unit.add and Unit.subtract operations). KELVIN_DIFFERENCE is unnecessary (as for all non-offset units).

**Comment by andi-huber** [comment]
> t1 = Quantities.create(5., Units.CELSIUS, ABSOLUTE);
> t2 = Quantities.create(3., Units.CELSIUS, ABSOLUTE);

Sorry, in case my question was badly phrased, to be more clear: I'd like to know what code a user has to write (utilizing UOM API) in order to get from given t1 and t2 to a result deltaT (t1-t2) which is a Quantity equivalent to
`Quantities.create(2., Units.CELSIUS, DIFFERENCE)`

I'm asking this, because it is not clear to me, whether we have any operators yet (other then the 'Quantities.create(...)' above), that would yield such a difference quantity as a calculation result.

**Comment by andi-huber** [comment]
@keilw It seems there is no use-case yet for LevelOfMeasurement.

To extend my question regarding the API design:

Given 
```
t1 = Quantities.getQuantity(5., Units.CELSIUS, ABSOLUTE);
t2 = Quantities.getQuantity(3., Units.CELSIUS, ABSOLUTE);
d = Quantities.getQuantity(1., Units.CELSIUS, DIFFERENCE);
```
Expected API behavior:
```
t1.subtract(t2) -> -271.15 °C (ABSOLUTE)
t1.add(d) -> 6 °C (ABSOLUTE)
d.add(d) -> 2 °C (DIFFERENCE)
d.multiply(2.) -> 2 °C (DIFFERENCE)
```
but what operators to use in order to calculate `t1 - t2` such that the result is `2 °C (DIFFERENCE)`

Correct me if I'm wrong, but right now we can only do this like so ...
```
deltaT = Quantities.getQuantity(
    t1.getValue().doubleValue() - t2.getValue().doubleValue(), 
Units.CELSIUS, DIFFERENCE);
```
Clearly that cannot be the solution we provide.

**Comment by desruisseaux** [comment]
@keilw: `ABSOLUTE` would indeed be the default for `QuantityFactory.create` methods, but my understanding of Andy's question is what should be the result of `Quantity` arithmetic operations, which involve other considerations. More on it below.

@dautelle: `Quantity` is more easily conceived as an absolute measurement for `Temperature` and `Mass` for instance, but more difficult to conceive as an absolute measurement for `Length` and `Time` for example. Most spatio-temporal measurements are relative to some position, and despite that we still qualify them as "quantity". If nevertheless we put the "absolute versus difference" flag in `Unit`, then the following code which was used to succeed in JSR-363 would fail in JSR-385 if _b_ is `Units.CELSIUS_DIFFERENCE`:

```
Unit<Temperature> a = Units.CELSIUS;
Unit<Temperature> b = ...;
UnitConverter c = a.getConverterTo(b);
```

Remember that `Unit` has two methods: `getConverterTo` which enforce compatible quantity at compile time and throw no checked exception, and `getConverterToAny` which does not enforce anything at compile time and throws the checked `IncommensurableException` if quantities are not compatible. We lost this compile-time safety if we introduce units like `Units.CELCIUS_DIFFERENCE`. Another consequence if that users can not rely anymore on the assumption that if _A_ and _B_ are in the same unit, then _A - B_ is in the same unit.

@andi-huber: intuitively, with `t1` and `t2` as `ABSOLUTE` quantities, the behavior of `t1 - t2` should be to produce a `DIFFERENCE` quantities. However I agree that it introduces an apparent inconsistency with the requirement that `t1.subtract(t2)` is equal to `t1.add(t2.multiply(-1))`. By symmetry, we should also require that `t1.add(t2)` is equal to `t1.subtract(t2.multiply(-1))` and we can imagine yet more combinations. The only way I can see at this time to get fully consistent behavior would be to state the following rules:

* `t1 + t2` -> `ABSOLUTE` if `t1` and `t2` are of the same sign, `DIFFERENCE` if they are of opposite sign.
* `t1 - t2` -> `DIFFERENCE` if `t1` and `t2` are of the same sign, `ABSOLUTE` if they are of opposite sign.

If at least one of `t1` or `t2` is `DIFFERENCE`, then the rules are as you described in your previous comment. It may look odd that all operations depend only on the "absolute versus difference" flag except the `ABSOLUTE` ± `ABSOLUTE` operation which depends also on the sign of the values, but it may be because having two flag values ("absolute" or "difference") is not as accurate as it could be (does "negative absolute value" really make sense?). However I don't think it is worth to complicate the API with more flags than "absolute" and "difference", so having to look at the sign of absolute values may be a good compromise.

**Comment by dautelle** [comment]
Length, Time and Temperature are absolute measurement/quantity (in any non-accelerated referential according to relativity). But is a temperature increment (e.g. difference between two absolute temperatures) of the same 'kind' as an absolute temperature? I don't think so. Hence different units and no converter between both (despite they have the same dimensions).

**Comment by andi-huber** [comment]
@desruisseaux instead of making the quantity's sign relevant, can we not have a 'consistent' set of rules? (see below)

Regardless of where we store the information ABSOLUTE/DIFFERENCE (either in Unit or Quantity), we also need to get arithmetic rules right! 

1) Quantity operations on eg. quantities of Unit.CELSIUS

Assuming commutativity of addition and multiplication: 

`ABSOLUTE` ± `ABSOLUTE` -> `ABSOLUTE`
`ABSOLUTE` ± `DIFFERENCE` -> `ABSOLUTE`
`DIFFERENCE` ± `DIFFERENCE` -> `DIFFERENCE`

`ABSOLUTE` * scalar -> `ABSOLUTE`
`DIFFERENCE` * scalar -> `DIFFERENCE`

`ABSOLUTE` * `ABSOLUTE` -> yields unit °C² ... ambiguous, what to do?
`DIFFERENCE` * `DIFFERENCE` -> yields unit °C² ... ambiguous, what to do?
`ABSOLUTE` * `DIFFERENCE` -> yields unit °C² ... ambiguous, what to do?

`ABSOLUTE` / `ABSOLUTE` -> scalar
`DIFFERENCE` / `DIFFERENCE` -> scalar

`ABSOLUTE` / `DIFFERENCE` -> NaN
`DIFFERENCE` / `ABSOLUTE` -> NaN

The API could be extended to also provide for the special case 
delta( `ABSOLUTE`  ,  `ABSOLUTE`) -> `DIFFERENCE` 
eg. by introducing a new method ...
```
t1 = Quantities.getQuantity(5., Units.CELSIUS, ABSOLUTE);
t2 = Quantities.getQuantity(3., Units.CELSIUS, ABSOLUTE);
deltaT = Quantities.getAsDifference(t1, t2); // new API !!!
```

2) Unit Conversion

`ABSOLUTE`.getConverterTo(`ABSOLUTE`) -> allowed
`DIFFERENCE`.getConverterTo(`DIFFERENCE`) -> allowed

`ABSOLUTE`.getConverterTo(`DIFFERENCE`) -> what to do?
`DIFFERENCE`.getConverterTo(`ABSOLUTE`) -> what to do?

---
I'm not sure if that's all rules, but we should be aware that we need to specify these.

**Comment by keilw** [comment]
What would `scalar` be, a number or something else? The idea of `deltaT = Quantities.getAsDifference(t1, t2)`; could be a little tricky because the API should be able to mirror it somewhere.

`Quantities` is a convenience facade, but all its methods are either used by SPI elements like `QuantityFactory` or wrap calls to` QuantityFormat.parse()`.

**Comment by andi-huber** [comment]
> 
> 
> What would `scalar` be, a number or something else? 

Yes a number, or more generally speaking a dimensionless quantity.

>The idea of `deltaT = Quantities.getAsDifference(t1, t2)`; could be a little tricky because the API should >be able to mirror it somewhere.
> 

Clearly I'm lacking insight as to where to put such an extension, but point being, we should consider extending the API, rather than having rules which involve the sign of quantities.

**Comment by keilw** [comment]
So if we assume two `ABSOLUTE` quantities like `10 °C / 3 °C` it would result in a number or `Dimensionless Quantity`?

Any sort of rule could not be restricted to `CELSIUS`, Other unit system extension modules like `uom-systems-*` etc. shall be able to add `FAHRENHEIT` or others like `Reaumur` although it is a bit historical, but it does not matter. Although it is nearly impossible to create an independent implementation of the `java.time` "low level" API implementing small parts like Chronology seems easier. And one could add ancient calendar systems like the Maya calendar, an Ancient Egyptian or Babylonian calendar for certain purposes. The same way it should be possible to add further units to a pre-defined rule set that may only contain `CELSIUS` out of the box in the RI.

**Comment by desruisseaux** [comment]
@dautelle: given that lengths are always the distance between two points and that there is no natural zero spatio-temporal position (ignoring big bang), we could debate about whether lengths are absolute measurements, regardless if classical physics or relativity. But anyway this is not this topic. A question for this JSR topic would be: why a temperature increment should be considered different than an absolute temperature, but a mass increment should not be considered different than an absolute mass?

@andi-huber: I agree with the rules you listed with 3 dampers:

* `ABSOLUTE - ABSOLUTE` -> `ABSOLUTE` seems surprising. It would force us to express 5°C - 3°C = -271.15 °C because the result must be equivalent to 2 K.
* I need to think more if I can find a use case for `ABSOLUTE / DIFFERENCE`.
* The question about `Unit.getConverterTo` behavior vanishes if we put the flag in `Quantity` instead of `Unit`.

I suggest to focus on `ABSOLUTE ± ABSOLUTE` for now, which seems to be the core of the issue we are trying to solve. I think the 2 other points are relatively minor in comparison. I will try to elaborate more on my sign proposal in the next days, and we can explore alternatives in parallel to see what fit best.

**Comment by keilw** [comment]
We could have both on Quantity, but I would not spawn off two enums that are more confusing. Either scrap `LevelOfMeasurement ` entirely and invent something "Absoludity", mayb not, I just can't think of anything other than `isAbsolute() vs. isRelative() or maybe isDifference()`.  
Very vaguely comparable to say `BigDecimal` having attributes like `scale` and `precision`. The `LevelOfMeasurement` is also called "scale" in many cases, but I would stick to "level" if we keep the enum at all?

**Comment by andi-huber** [comment]
Regarding isAbsolute() vs. isRelative() or maybe isDifference():

I suggest to use an enum like this to 'put' on `Quantity` or `Unit`:
```
public enum ArithmeticCharacteristics {

  /** no arithmetic distinction is made between absolute or difference on quantities having 
   * UNIFIED characteristics, eg. any quantity of unit KELVIN
   **/
  UNIFIED,
  
  /** special arithmetic rules apply to quantities having ABSOLUTE characteristics
   * eg. an absolute quantity of unit CELSIUS
   **/
  ABSOLUTE,
  
  /** special arithmetic rules apply to quantities having DIFFERENCE characteristics
   * eg. a difference quantity of unit CELSIUS
   */
  DIFFERENCE
}
```
Idea being, that for most of our Quantities/Units this characteristics is `UNIFIED` and can not be something else, like for eg `KELVIN`. While for `CELSIUS` it can either be `ABSOLUTE` or `DIFFERENCE` but never `UNIFIED`.

My thoughts on solving

> ABSOLUTE * ABSOLUTE -> yields unit °C² ... ambiguous, what to do?
> DIFFERENCE * DIFFERENCE -> yields unit °C² ... ambiguous, what to do?
> ABSOLUTE * DIFFERENCE -> yields unit °C² ... ambiguous, what to do?
> 

Lets say we have 2 quantities `a` and `b` then we can have arithmetic rules to resolve the ambiguity problem for multiplication by (pseudo-code)
```
a * b -> convertToUnified(a) * convertToUnified(b) -> yields unit K²
```
provided, we do have rules for converting non-unified to unified,  example (pseudo-code):
```
a = Quantities.getQuantity(5., Units.CELSIUS, ABSOLUTE);
b = Quantities.getQuantity(1., Units.CELSIUS, DIFFERENCE);

a * b  -> toKelvin(a) * toKelvin(b) -> yields unit K² ... no longer ambiguous
```
I conclude we need to allow for quantities of CELSIUS (either difference or absolute) to be converted to KELVIN (always unified). We need to specify how to do this generally (not only Temperature).

**Comment by andi-huber** [comment]
I created a wiki page summarizing the rules at Indriya [1]. Feel free to copy/move it over to here. I don't have editing rights on this wiki.

[1] https://github.com/unitsofmeasurement/indriya/wiki/Arithmetic-Rules-for-Difference-vs-Absolute-Quantities-(DRAFT)

**Comment by desruisseaux** [comment]
Thanks Andy for the wiki page. I think it is really useful. I will try to copy to unit-API during the weekend and elaborate on alternatives regarding `ABSOLUTE - ABSOLUTE` (sign comparison versus fixing the result to `ABSOLUTE`).

**Comment by keilw** [comment]
If it is just absolute vs. relative/difference, I would leave a boolean flag. It feels natural with the `UnitConverter.isLinear()` example also mentioned here ;-)
Since the other thread was already resolved (and the majority voted for using `Quantity`) @andi-huber @dautelle since you were the most active/only participants other than @desruisseaux with arguments like 
> LevelOfMeasurement fits well in Unit but is not exactly what we need for 

or

> I would suggest not changing current units definition ( "scaled dimensions") 

How do you feel about `LevelOfMeasurement` either in `Unit`, `Quantity` (where a majority of all participants found it would be better) or completely drop it because an absolute/different attribute in `Quantity` is totally sufficient for our current needs?

I would not oppose refactoring it into `Unit`, but there it should be a constant unique to each `Unit`, so in theory we could have a `KELVIN_INTERVAL` or `METRE_INTERVAL` somewhere, but the units defined in `Units` or other places have one distinct immutable `LevelOfMeasurement`.

**Comment by andi-huber** [comment]
@keilw I suggest to drop `LevelOfMeasurement`, simply because we don't use it. Why worry about introducing a feature, which has no evident use-case within Indriya. We can always reintroduce it later, should we find a use-case. Also please note that I'm currently suggesting [1] an `enum` instead of a `boolean` regarding absolute vs different.

[1] https://github.com/unitsofmeasurement/indriya/wiki/Arithmetic-Rules-for-Difference-vs-Absolute-Quantities-(DRAFT)

**Comment by keilw** [comment]
@andi-huber If there really can be more than one even an "Unknown" value, why not, but what about simply calling it [Scale](https://en.wikipedia.org/wiki/Absolute_scale)?

**Comment by andi-huber** [comment]
@keilw My thought process from early morning, following the idea of having 3 instead of 2 values for this arithmetic characteristics, is not set in stone at all. As Martin has offered, he'll think this through next weekend. I'll also think this through again. If it makes sense to stick with my proposal of 3 values ABSOLUTE, DIFFERENCE and UNIFIED then I'd rather stick with the term 'Characteristics'. If it terns out that only 2 values ABSOLUTE, DIFFERENCE are needed, the term 'Scale' seems a good choice.

**Comment by keilw** [comment]
Either way I think we should keep it as a static inner enum with the affected type (`Quantity` for all we know). And if we ever find a need for `LevelOfMeasurement`, we can recover it. Indriya has at least one or two research branches for CompoundUnit, etc. I think I'll take a branch of the API simply to preserve a few discussion items. and then at least `@deprecate` `LevelOfMeasurement `so it can be removed if no further use is found for it.

**Comment by keilw** [comment]
Sure, think about it over the weekend. I deprecated `LevelOfMeasurement`, not yet removing it from `master`. If there is no use case, then it could be removed in favor of another definition. Tomorrow the CGPM in Versailles will decide on the redefined SI units. Hope it goes through, although it has no direct impact on this issue. Even in the unlikely event they rejected or postponed it again we shall add such new features or improvements. At most a MR or future release may have to deal with more SI changes if it did not happen by May 20 next year.

**Comment by andi-huber** [comment]
In the meantime I've extended the wiki page [1], which now draws a complete picture of rules and API extensions required. Again its just a draft! Also the naming is open for discussion.

Few things I've learned:
1) Units might need a new property which tells us whether the unit is 'allowed' to have a distinction 'absolute/difference' or not, handled by an enum `ArithmeticPolicy`.
2) With all the Units currently defined in Indryia, there's only a single one, that requires a policy other than `UNIFIED_ONLY` which is CELSIUS.
3) There is a way, to have new enum `ArithmeticCharacteristic` as an implementation detail, not exposed to the public API, see this chapter [2]

[1] https://github.com/unitsofmeasurement/indriya/wiki/Arithmetic-Rules-for-Difference-vs-Absolute-Quantities-(DRAFT)
[2] https://github.com/unitsofmeasurement/indriya/wiki/Arithmetic-Rules-for-Difference-vs-Absolute-Quantities-(DRAFT)#api-extension-quantity-creation-2

**Comment by andi-huber** [comment]
... I need to add, that with a bit more 'thinking this through' over the weekend, we find a way to simplify the rule-set, such that we don't need to distinction between UNIFIED and ABSOLUTE. In other words, currently we have A, D, U operands for the rule-set, maybe its sufficient to only have A, D (with A=U).

**Comment by keilw** [comment]
If you would like to give that a try over the weekend, a boolean flag (now also manifested in `AbstractQuantity` and relevant subclasses) is there all the time. So if 2 are enough A would be `true`, anything else would be `false`. If there is a realistic chance we'll ever need one or the other different choice, we could turn it into an `enum`, otherwise the flag might just do.

**Comment by desruisseaux** [comment]
Hello Andi. There is some first remarks:

It seems to me that there is a one-to-one relationship between `ArithmeticPolicy.UNIFIED_ONLY` and `LevelOfMeasurement.RATIO`, and between `ArithmeticPolicy.ABSOLUTE_OR_DIFFERENCE` and `LevelOfMeasurement.INTERVAL`. Is that right?

One clarification: in the wiki page, are _A_ and _D_ always Quantity of the same type? I ask because expressions like `A / D -> scalar` implies so (otherwise the result would not be a scalar).

I concur with previous comment, that we may not need `ArithmeticCharacteristic.UNIFIED`. For any unit of kind `LevelOfMeasurement.RATIO` (or `ArithmeticPolicy.UNIFIED_ONLY`), we can ignore the absolute versus difference flag and just perform all computations in the same way. It also implies that `Quantities.getDifferenceQuantity(5, KELVIN)` does not need to throw an exception I think, since it seems legitimate to create a quantity representing a difference of 5 K and use it in calculations.

Before to move to elaborate on comments, do we agree to produce a page showing the arithmetic rules with only "absolute" and "difference" flags? If yes, do we edit the current one or do we create and modify a copy, so the two alternatives can be compared?

**Comment by keilw** [comment]
Guys, could we have just one, at most two enums please. Otherwise not only the JAR size would again (even without Maven metadata) cross 50kB very soon, it also became very complex and error-prone with all the possible combinations.

Those who voted on having `LevelOfMeasurement` on a `Quantity` earlier (everyone of myself, @dautelle, @desruisseaux and probably @andi-huber) does everyone agree, it is still needed and should be a constant on `Unit` now?  

+1 on trying to use an "absolute/difference" flag only, whether it's on Quantity or Unit, whatever works better. If the LevelOfMeasurement enum helps for additional decision making, then why not keep it and apply it where it's found to be of best use. 

The wiki pages don't have a such a strong history as the actual Git codebase, but there are revisions, thus please use the existing wiki page, otherwise it is almost impossible co compare.

**Comment by andi-huber** [comment]
> It seems to me that there is a one-to-one relationship between `ArithmeticPolicy.UNIFIED_ONLY` and `LevelOfMeasurement.RATIO`, and between `ArithmeticPolicy.ABSOLUTE_OR_DIFFERENCE` and `LevelOfMeasurement.INTERVAL`. Is that right?
>
I was not sure about this, but seems plausible. I trust your insight.

> One clarification: in the wiki page, are _A_ and _D_ always Quantity of the same type? I ask because expressions like `A / D -> scalar` implies so (otherwise the result would not be a scalar).
> 
YES, the wiki page exemplifies the rules using Unit `Temperature` and does not mix types.

> I concur with previous comment, that we may not need `ArithmeticCharacteristic.UNIFIED`. For any unit > of kind `LevelOfMeasurement.RATIO` (or `ArithmeticPolicy.UNIFIED_ONLY`), we can ignore the absolute > versus difference flag and just perform all computations in the same way.

Agreed

> It also implies that `Quantities.getDifferenceQuantity(5, KELVIN)` does not need to throw an exception I > think, since it seems legitimate to create a quantity representing a difference of 5 K and use it in
> calculations.
> 
I'm not sure in this case how to tell the user, that use of `Quantities.getDifferenceQuantity(5, KELVIN)` and  `Quantities.getAbsoluteQuantity(5, KELVIN)` are the same, we don't care. Both are treated with UNIFIED (=RATIO) calculus. Hence my idea to fail early, and tell the user he might have a misconception. Allowing or disallowing this, both works for me.

> Before to move to elaborate on comments, do we agree to produce a page showing the arithmetic rules with only "absolute" and "difference" flags? If yes, do we edit the current one or do we create and modify a copy, so the two alternatives can be compared?

Good question, the safe way is to make a wiki-page copy, showcasing the A, D only rules.

**Comment by keilw** [comment]
Once a decision is made, any redundant or outdated (because the API/impl won't match any more) Wiki page should go away, so of course we could spawn of multiple pages, but they should not be around for longer than the Proposed Final Draft. I'm a bit uncomfortable with too many variations like `Quantities.getDifferenceQuantity(5, KELVIN)` vs. `Quantities.getAbsoluteQuantity(5, KELVIN)`. 
What should happen with `Quantities.getQuantity(5, KELVIN)` ? And it must be possible to also create it via `QuantityFactory`, Even for the RI it would be dangerous and extremely dodgy to suddenly drop exiting calls like `Quantities.getQuantity(5, KELVIN)`, in the API it is impossible to break QuantityFactory, so please think of what is the best and least intrusive way of creating them also using the API.

**Comment by andi-huber** [comment]
@keilw I think first of all, we need to work on the rule-set, to find out what our minimum requirements are. What I have in mind with the A,D,U proposal and its wiki-page is to have a more concrete way of discussing this issue. Also for now I would not care too much about naming and compiled size. I'm sure we'll progress to a point soon, where we converge to a solution and find what these minimum requirements are. Only then I would suggest to take care of naming and sizing issues. I agree, once we have done all that, only one concise wiki page shall remain.

**Comment by desruisseaux** [comment]
@keilw: I agree with Andi; lets get the rules right for now, even if it results in a verbose API. After we reached that state, then we can explore ways to trim the API side down without breaking the rules. It is a little bit like writing a complicated mathematical equation at first, verify that it works, then try to simplify it.

I will clone the wiki page with a variant containing only the A and D proposals. I will try to do that tomorrow (Tuesday). Later one, we can merge the various wiki page in a single one summarizing our conclusion.

**Comment by keilw** [comment]
I don't think it has an impact on this, but the `Kelvin` will also be redefined ;-)

**Comment by filipvanlaenen** [comment]
I've been thinking about this issue for a couple of days now, and I still have problems understanding how things should work and behave.

One thing I've started to realize though is that arithmetic with temperatures only become a problem if one chooses to express them in °C or °F, and not if they're expressed in K. So I think that the units are the problem, but I don't think we need a flag to mark a unit as problematic. If a unit converts in a strictly multiplicative way from the base unit, everything is fine. If anything else is involved (addition, logarithm, exponentiation, hyperbolic sine, or whatever), things don't work out.

Also, we should take care that we don't make a solution that works for additional conversions only.

I wonder therefore whether the distinction really is as follows
1. Arithmetic has to be done after conversion to the base unit
2. Arithmetic has to be done on the nominal values

For the first case, one should try to convert the result back to the original units, but that may not always be possible (e.g. how to convert K² to °C²?).

For the second case, all the operands have to be expressed in the same unit (or with strictly multiplicative conversions), or we throw an exception. This means that you can add °C with °C, but not °C with °F. If someone really needs to be able to calculate 1 °C + 1 °F, they can create their own library with their own logic, but I think this is out of scope for a reference implementation.

As for the API, I think this means we should perhaps rather overload the `add(operand)` method with `add(operand, mode)`. The `add(operand)` method throws an exception if one of the units has a non-multiplicative conversion to the base unit, whereas the `add(operand, mode)` method throws an exception if the units don't convert internally in a strictly multiplicative manner.

**Comment by andi-huber** [comment]
> 
> 
> I've been thinking about this issue for a couple of days now, and I still have problems understanding how things should work and behave.
> 
Thanks Filip, in case you are not aware, to make any progress with this discussion, we first of all should get the arithmetic rule-set [1] right, then simplify it as much as possible and only then worry about how to design the API in order to implement the rules.

> One thing I've started to realize though is that arithmetics with temperatures only become a problem if one chooses to express them in °C or °F, and not if they're expressed in K. So I think that the units are the problem, but I don't think we need a flag to mark a unit as problematic. If a unit converts in a strictly multiplicative way from the base unit, everything is fine. If anything else is involved (addition, logarithm, exponentiation, hyperbolic sine, or whatever), things don't work out.
> 
Agreed! In that sense (and as Martin suggests) 'problematic' Units correspond to not being of type RATIO (=UNIFIED). You are suggesting, that we don't need to define this explicitly, because we already know it implicitly. Seems plausible to me, however this is something, we can worry about, once we have the rule-set.

> Also, we should take care that we don't make a solution that works for additional conversions only.
> 
> I wonder therefore whether the distinction really is as follows
> 
>     1. Arithmetic has to be done after conversion to the base unit
> 
>     2. Arithmetic has to be done on the nominal values
> 
> 
> For the first case, one should try to convert the result back to the original units, but that may not always be possible (e.g. how to convert K² to °C²?). People will have to live with that 1 °C × 1 °C is 75,158.2225 K².
> 
Agreed! That's in line with what my rule-set proposal suggests.  
`A * A -> U`  
`A * D -> U`  
`D * A -> U`

> For the second case, all the operands have to be expressed in the same unit (or with strictly multiplicative conversions), or we throw an exception. This means that you can add °C with °C, but not °C with °F. If someone really needs to be able to calculate 1 °C + 1 °F, they can create their own library with their own logic, but I think this is out of scope for a reference implementation.
> 
Disagreed! As long as every(!) Quantity of CELSIUS or FAHRENHEIT does indicate whether it is of type ABSOLUTE or DIFFERENCE, then we can make use of the rule-set:  
eg. 1 °C + 1 °F is well defined if we know for both operands what they are (absolute ore difference)

> As for the API, I think this means we should perhaps rather overload the `add(operand)` method with `add(operand, mode)`. The `add(operand)` method throws an exception if one of the units has a non-multiplicative conversion to the base unit, whereas the `add(operand, mode)` method throws an exception if the units don't convert internally in a strictly multiplicative manner.
>
Not required, according to prev. comment. Again, we can worry about the 'how extend the API', once we agree on the arithmetic rule-set.

[1] https://github.com/unitsofmeasurement/indriya/wiki/Arithmetic-Rules-for-Difference-vs-Absolute-Quantities-(DRAFT)

**Comment by keilw** [comment]
@filipvanlaenen @andi-huber Thanks for continuing the discussion with a few more eyes and thoughts to go over this.

As long as most of this, especially 

1. Arithmetic has to be done after conversion to the base unit
2. Arithmetic has to be done on the nominal values

Is restricted to e.g. any value/unit without `UNIFIED_ONLY` this sounds OK and reasonable, because we don't want nasty side-effects e.g. when adding `5 inch` to a `foot`. 
As it's mentioned in the Wiki I don't really see an easy way for compile-time checking in those cases. Extending `Quantity` (or the likes of `Temperature` in this case) was brushed by @desruisseaux but I think it would cause a big mess and not really help compared to one or the other attribute for such rules. 
I would not throw an `IllegalArgumentException` btw. but rather a `MeasurementException` or if we really find it necessary another subclass of it. If we really stuck to  the JDK then `ArithmeticException` could work in that case.

**Comment by desruisseaux** [comment]
I have [created a new wiki page](https://github.com/unitsofmeasurement/unit-api/wiki/Arithmetic-rules-for-Difference-versus-Absolute-quantities) adapted from [Andi's page on Indriya](https://github.com/unitsofmeasurement/indriya/wiki/Arithmetic-Rules-for-Difference-vs-Absolute-Quantities-(DRAFT)). Differences are:

* `UNIFIED` characteristic omitted.
* `ArithmeticPolicy` replaced by `LevelOfMeasurement`.
* `ArithmeticCharacteristic` renamed `MeasurementType` because I think that this enumeration does not really specify the behavior of arithmetic operations, but rather how values are converted to system units. The golden rule is that all arithmetic operations (no matter if the operands are absolute or incremental) shall behave as if the values were converted to system unit before calculation. The "absolute" versus "difference" characteristics modifies the way this conversion is done, rather than the way the arithmetic operations are done after conversion.
* `DIFFERENCE` renamed `INCREMENTAL`. I found "incremental" used on [this site](http://www.bogen-electronic.com/en/magnetic-measurement-solutions/technology/absolute-and-incremental-measurement.html) and it seems to fit well our purpose. At least, it carries better the meaning of the sign (a negative value is a decrement).

However I suggest to not focus on the naming for the early stage of review. The initial focus should be on ensuring that the arithmetic ground is solid. After that we can look at aspects like naming.

I abandoned my idea of using the sign of operands for determining if the result of "absolute ± absolute" should be absolute or incremental, because when trying to write down examples I found cases in contradictions with arithmetic laws in a way that I could not fix. Instead I adopted Andi's proposal of an additional method in `Quantity` interface.

@filipvanlaenen indeed some operations on °C or °F are ambiguous (i.e. depend if the quantity is an increment or not) while operations on K are non-ambiguous. But this is not specific to temperature units. More generally, all operations involving units of `INTERVAL` level of measurement are potentially ambiguous while the same operations involving units of `RATIO` level of measurement are non-ambiguous. The wiki pages are proposals trying to resolve the ambiguity.

**Comment by keilw** [comment]
Sounds like it's getting a little more concrete. I am not completely against 
`QuantityFactory.createIncrement(Number, Unit)` creates an `INCREMENTAL` quantity.
It makes it a little more complicated, but e.g. JSON-P also has a BuilderFactory with createJsonBuilder vs. createJsonArrayBuilder() so it might be OK.

`Quantity.difference(Quantity)` as discussed above.
I don't like `UnitConverter.deltaConvert(double)`  because it adds yet another term for the same thing. It would be utterly confusing, plus the convention of above proposal in `QuantityFactory` would make it better to have `UnitConverter.convertIncrement( ` or `UnitConverter.convertDifference()` but please let us stick to either one or at most two words not three.

**Comment by desruisseaux** [comment]
Hello Andy. Thanks for the review. For question 1 I mean the following situation:

```
A = 2 cm
B = 1 inch
A + B = 3.54 cm (B has been converted in the unit of A).
B + A = 1.7874… inch (A has been converted in the unit of B).
```

I think that we do not need to require A + B and B + A to return a result in the same unit of measurement, as long as the two results are equivalent (0.0354 m in above example). The reason is that different units will not make any calculation wrong (ignoring rounding errors) since they describe the same length. This is different than the situation where we had (prior #95 discussion)  1°C + 2 K giving -271.15°C and 2 K + 1°C giving 276.15 K, since in that case those two results are not just expressing the same temperature in different units; they are truly different results.

The proposed rule is that all calculations must be performed as if computed in system units. If, after that, the implementation which to convert the result to any unit of its choice, it is okay since if that result is used in another computation, the very first step will be to convert it back again to the system unit, i.e. back to the original result prior conversion to an arbitrary unit.

For question 2, the need for `UnitConverter.deltaConvert(double)` is because a critical step in the proposal is the conversion to system unit. The difference between `ABSOLUTE` and `INCREMENTAL` are not in the arithmetic, but in the way this conversion is done. For example in `1°C + 2 m°C` where 1°C is `ABSOLUTE`, the calculation is performed as below:

* 1°C is converted to 274.15 K using `UnitConverter.convert(double)`.
* For the 2 m°C operand, there is a choice:
    * If 2 m°C is `ABSOLUTE`, then it is converted to 273.152 K using `UnitConverter.convert(double)`.
    * If 2 m°C is `INCREMENTAL`, then it needs to be converted to 0.002 K. The `UnitConverter.convert(double)` does not perform that job. We need a conversion applying only the scale factor, without offset. This is where `UnitConverter.deltaConvert(double)` come into play.
* The final result can then be converted to °C or something else, at implementation choice.

Note that the proposed `UnitConverter.deltaConvert(double)` method is inspired by JavaSE `AffineTransform.deltaTransform(Point2D)` method [(javadoc)](https://docs.oracle.com/en/java/javase/11/docs/api/java.desktop/java/awt/geom/AffineTransform.html#deltaTransform(java.awt.geom.Point2D,java.awt.geom.Point2D)). This is the same semantic, but in one dimension instead than two dimensions.

**Comment by keilw** [comment]
Having a JDK method (although AWT could soon be removed from Java SE as well) used for inspiration sounds OK. If we have a name for that flag/enum and `QuantityFactory.create()` only using the optional extra argument, then there would be less terms all across the place, so `deltaConvert` might be OK in that case.

**Comment by andi-huber** [comment]
It all makes sense to me now. Also the example you showed on the wiki page following quote 1) already appeared clear to me, I just was not sure if there is more to it.

Thanks for clarifying!

**Comment by keilw** [comment]
To avoid confusing it with the actual type of the `Quantity` (like `Mass` or `Length`) could we call the distinction of measurement (former `ArithmeticCharacteristic`) `MeasurementKind` instead of `MeasurementType`?

**Comment by dautelle** [comment]
Sorry, but I hate all generic word which do not bring much informative value. In this particular case, I would go with MeasurementScale (e.g. Absolute Scale, Incremental Scale). Other words would apply to almost anything and for measurement don't say much.

**Comment by andi-huber** [comment]
As Martin managed to work out in great detail here [Wiki Page](https://github.com/unitsofmeasurement/unit-api/wiki/Arithmetic-rules-for-Difference-versus-Absolute-quantities), we do have 2 enums to distinguish and not to confuse with one another. One is a property associated with `Unit`, the other is a property associated with `Quantity`, both at first glance seem similar, but do serve a different purpose within the API.

1) Property of `Unit` ... `LevelOfMeasurement (NOMINAL, ORDINAL, INTERVAL, RATIO)`
2) Property of `Quantity` ... `MeasurementType (ABSOLUTE, INCREMENTAL)`

I guess the challenge now is to find a good terminology, that also accounts for the distinct purpose of the two enums above. 

I have a felling, that the term `Scale` might suggest association with `LevelOfMeasurement`,  which we should(?) avoid. I don't know.

**Comment by dautelle** [comment]
Since an "incremental quantity" should be stated using an interval unit (LevelOfMeasurement). Do we need a property  "MeasurementType" on quantity?

**Comment by desruisseaux** [comment]
An incremental quantity can be stated with both `INTERVAL` and `RATIO` units. For example we can have an increment of 5°C or 5 K. As Andy said, those two enumerations are related but not redundant. `LevelOfMeasurement` defines the arithmetic operations than we can apply between quantities, approximately as shown in [Wikipedia table](https://en.wikipedia.org/wiki/Level_of_measurement#Comparison) (I said "approximately" because keeping `+` consistent with `-` is tricky, which has leaded us to propose a new `Quantity.difference(Quantity)` method). When computing 2×3°C, the enumerations are used as below:

1. The `INTERVAL` level of measurement associated to the °C units tells us that we can not apply multiplications or divisions on quantities in that unit.
2. We need to convert the quantities to a unit having `RATIO` level of measurement (namely Kelvin degrees) so we can apply the multiplication.
3. Only after we have determined that such conversion is necessary - that not doing it would produce different numerical results - only then the "absolute" versus "incremental" characteristics tell us how to do the conversion from the `INTERVAL` to the `RATIO` unit.

We can convert an `ABSOLUTE` quantity from `INTERVAL` unit to `RATIO` unit (e.g. 3°C converted to 275.15 K), but we can also convert an `INCREMENTAL` quantity from `INTERVAL` unit to `RATIO` unit (e.g. 3°C converted to 3 K).

It may be possible to use the same `LevelOfMeasurement` enumeration for those two different purposes, but we would have to check if we can do that in a way to do not introduce confusion.

**Comment by desruisseaux** [comment]
However, as said previously, a `LevelOfMeasurement` enumeration on `Units` is technically not strictly required. The only information which is mandatory (as far as I can see) is the "absolute" versus "incremental" aspect of `Quantity`. So whether keeping `LevelOfMeasurement` would bring more clarity or confusion is an open question.

**Comment by keilw** [comment]
If only the absolute/relative `enum` (or flag) is required, then I would remove `LevelOfMeasurement` for now. It would help data analysis, but unless we really have a use case e.g. with BigData, monitoring, etc. that would greatly benefit, we could still add it later with a MR or update release.

**Comment by keilw** [comment]
I hope, some of us are able to start implementing what's necessary also over Christmas? We should file a Public Review no later than first thing in January. @filipvanlaenen and I have a talk at OOP in Munich late January, so it would be great to have the Public Review ongoing by then. Given the SI redefinition milestone, PFD (practically final) and a Final Release should take place in the first half of Q2 at the latest. There are only 2 EC F2F meetings in 2019, one is scheduled on May 15 in Tokyo: https://jcp.org/en/whatsnew/calendar2019. Which would be a great opportunity to report to the EC after the JSR went final or immediately before (The SI does on May 20, 2019)

**Comment by keilw** [comment]
@andi-huber @dautelle I mentioned earlier, that `MeasurementType` can be too easily confused with the actual (generic) type of the `Quantity`, so I suggested to call it `MeasurementKind` which @desruisseaux said, he would also be fine with. [LevelOfMeasurement](https://en.wikipedia.org/wiki/Level_of_measurement) is often called "Scale of Measure" or "Scale of Measurement", so `MeasurementScale` IMO is out of question. It would be totally confusing with what we currently called `LevelOfMeasurement`, even if we dropped one of them.

**Comment by dautelle** [comment]
Found on the web: "The clock is an absolute measurement system, it will tell you a point in time. A stop watch is an incremental system, it will tell you how many seconds (increments) have gone by since the start of the measurement. " - What about MeasurementSystem ?

**Comment by keilw** [comment]
IMO it sounds to close to `SystemOfUnits` which some sources also call [SystemOfMeasurement](https://en.wikipedia.org/wiki/System_of_measurement).

**Comment by keilw** [comment]
If this goes strictly into `Quantity`, why not define a `public static enum` there like `Scale`?
See https://en.wikipedia.org/wiki/Absolute_scale The term "scale" sounds reasonable, but I would rather not call it MeasurementScale, because it can be more easily confused with https://en.wikipedia.org/wiki/Level_of_measurement

**Comment by keilw** [comment]
I added a static `enum` `Scale` to `Quantity`. Please have a look. It's not applied yet, but the idea would be to replace the simple `boolean` flag with a `getScale()` method or similar. The names are based on the terms in Wikipedia, `ABSOLUT` and `RELATIVE`. There is no mention of aliases like "iterative", etc. No problem to mention any of these in JavaDoc, but I would stick to official terms here.

**Comment by keilw** [comment]
Hi, Hope we can do something with that enum over the holidays? Unless the LevelOfMeasurement was still of value for any operation, maybe we'll remove it for now until an actual need comes up?

**Comment by desruisseaux** [comment]
Hello Werner. I'm in a very big rush at work those time, including weekends. When I will start having some weekends free, I will implement the new feature in Seshat. Sorry for the delay…

**Comment by keilw** [comment]
Hi Martin. No worries, it is similar here (not on weekends yet;-) If nobody needs LevelOfMeasurement I'd remove it (still stashed away in a branch) and then all implementations can use the remaining enum.

**Comment by keilw** [comment]
Btw, LevelOfMeasurement was removed now. And all places that referred to a `boolean` `isAbsolute()` flag now use the `Quantity.Scale`. 
I added a placeholder in the spec: https://docs.google.com/document/d/12KhosAFriGCczBs6gwtJJDfg_QlANT92_lhxUWO2gCY/edit#heading=h.o50xwl22u27k. Please feel free to make suggestions or changes for that or the initial definition chapter on Quantity. It could be based on e.g. the Wiki page here, but not too lengthy. Of course if a concrete example for Temperature calculation could fit either under **Use Cases** or somewhere under **Definitions**, that sounds fine.
